crates/rust-analyzer/src/reload.rs

```diff
@@ -448,7 +462,19 @@ impl GlobalState {
         };
         let mut change = Change::new();
 
-        if same_workspaces {
+        // `self.fetch_proc_macros_queue.request_op(cause, proc_macro_paths)` is only called in
+        // when `switch_workspaces` is called _without_ changing workspaces. This typically occurs
+        // when build scripts have finishing building, but when rust-analyzer is used with a
+        // rust-project.json, the build scripts have already been built by the external build system
+        // that generated the `rust-project.json`.
+
+        // Therefore, in order to allow _new_ workspaces added via rust-project.json (e.g., after
+        // a workspace was already added), we check whether this is the same workspace _or_
+        // if any of the new workspaces is a `rust-project.json`.
+        //
+        // The else branch is used to provide better diagnostics to users while procedural macros
+        // are still being built.
+        if same_workspaces || different_workspaces.iter().any(|ws| ws.is_json()) {
```

Comment:
This also still doesn't fix the case where build scripts are disabled and rust-project.json are used without loading up workspaces after the start. I think I overcomplicated this part here, we should just replace this entire block of lines (477-490) with
```rs
if self.config.expand_proc_macros() {
    self.fetch_proc_macros_queue.request_op(cause, proc_macro_paths);
}
```
And set up the proc-macros errors when building up the crate graph, that is change https://github.com/rust-lang/rust-analyzer/blob/e7337fc9c3b4995fdab4e90871bfb48a26d5eaec/crates/base-db/src/input.rs#L20
to 
```rs
 pub type ProcMacroPaths = FxHashMap<CrateId, Result<(Option<String>, AbsPathBuf), String>>; 
 ```
 and fix up where we set those up.

Reply:
Currently we signal with `Option::None` that proc-macros haven't been build but this overcomplicates things by pushing the error generation to a later place, where we don't really have the information anymore.